File: rb/Irail_sdk.rb
# Irail SDK

require_relative 'utility/struct/voxgig_struct'
require_relative 'core/utility_type'
require_relative 'core/spec'
require_relative 'core/helpers'

# Load utility registration
require_relative 'utility/register'

# Load config and features
require_relative 'config'
require_relative 'feature/base_feature'
require_relative 'features'

# Load typed models (Struct value objects).
require_relative 'Irail_types'


class IrailSDK
  attr_accessor :mode, :features, :options

  def initialize(options = {})
    @mode = "live"
    @features = []
    @options = nil

    utility = IrailUtility.new
    @_utility = utility

    config = IrailConfig.make_config

    @_rootctx = utility.make_context.call({
      "client" => self,
      "utility" => utility,
      "config" => config,
      "options" => options || {},
      "shared" => {},
    }, nil)

    @options = utility.make_options.call(@_rootctx)

    if VoxgigStruct.getpath(@options, "feature.test.active") == true
      @mode = "test"
    end

    @_rootctx.options = @options

    # Add features from config.
    feature_opts = IrailHelpers.to_map(VoxgigStruct.getprop(@options, "feature"))
    if feature_opts
      items = VoxgigStruct.items(feature_opts)
      if items
        items.each do |item|
          fname = item[0]
          fopts = IrailHelpers.to_map(item[1])
          if fopts && fopts["active"] == true
            utility.feature_add.call(@_rootctx, IrailFeatures.make_feature(fname))
          end
        end
      end
    end

    # Add extension features.
    extend_val = VoxgigStruct.getprop(@options, "extend")
    if extend_val.is_a?(Array)
      extend_val.each do |f|
        if f.respond_to?(:get_name)
          utility.feature_add.call(@_rootctx, f)
        end
      end
    end

    # Initialize features.
    @features.each do |f|
      utility.feature_init.call(@_rootctx, f)
    end

    utility.feature_hook.call(@_rootctx, "PostConstruct")
  end

  def options_map
    out = VoxgigStruct.clone(@options)
    out.is_a?(Hash) ? out : {}
  end

  def get_utility
    IrailUtility.copy(@_utility)
  end

  def get_root_ctx
    @_rootctx
  end

  def prepare(fetchargs = {})
    utility = @_utility
    fetchargs ||= {}

    ctrl = IrailHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "prepare",
      "ctrl" => ctrl,
    }, @_rootctx)

    opts = @options
    path = VoxgigStruct.getprop(fetchargs, "path") || ""
    path = "" unless path.is_a?(String)
    method_val = VoxgigStruct.getprop(fetchargs, "method") || "GET"
    method_val = "GET" unless method_val.is_a?(String)
    params = IrailHelpers.to_map(VoxgigStruct.getprop(fetchargs, "params")) || {}
    query = IrailHelpers.to_map(VoxgigStruct.getprop(fetchargs, "query")) || {}
    headers = utility.prepare_headers.call(ctx)

    base = VoxgigStruct.getprop(opts, "base") || ""
    base = "" unless base.is_a?(String)
    prefix = VoxgigStruct.getprop(opts, "prefix") || ""
    prefix = "" unless prefix.is_a?(String)
    suffix = VoxgigStruct.getprop(opts, "suffix") || ""
    suffix = "" unless suffix.is_a?(String)

    ctx.spec = IrailSpec.new({
      "base" => base, "prefix" => prefix, "suffix" => suffix,
      "path" => path, "method" => method_val,
      "params" => params, "query" => query, "headers" => headers,
      "body" => VoxgigStruct.getprop(fetchargs, "body"),
      "step" => "start",
    })

    # Merge user-provided headers.
    uh = VoxgigStruct.getprop(fetchargs, "headers")
    if uh.is_a?(Hash)
      uh.each { |k, v| ctx.spec.headers[k] = v }
    end

    _, err = utility.prepare_auth.call(ctx)
    raise err if err

    utility.make_fetch_def.call(ctx)
  end

  def direct(fetchargs = {})
    utility = @_utility

    # direct() is the raw-HTTP escape hatch: it always returns a result hash
    # ({ "ok" => ..., ... }) and never raises. prepare() raises on error, so
    # trap that and surface it in the hash.
    begin
      fetchdef = prepare(fetchargs)
    rescue IrailError => err
      return { "ok" => false, "err" => err }
    end

    fetchargs ||= {}
    ctrl = IrailHelpers.to_map(VoxgigStruct.getprop(fetchargs, "ctrl")) || {}

    ctx = utility.make_context.call({
      "opname" => "direct",
      "ctrl" => ctrl,
    }, @_rootctx)

    url = fetchdef["url"] || ""
    fetched, fetch_err = utility.fetcher.call(ctx, url, fetchdef)

    return { "ok" => false, "err" => fetch_err } if fetch_err

    if fetched.nil?
      return {
        "ok" => false,
        "err" => ctx.make_error("direct_no_response", "response: undefined"),
      }
    end

    if fetched.is_a?(Hash)
      status = IrailHelpers.to_int(VoxgigStruct.getprop(fetched, "status"))
      headers = VoxgigStruct.getprop(fetched, "headers") || {}

      # No-body responses (204, 304) and explicit zero content-length must
      # skip JSON parsing — calling json() on an empty body errors.
      content_length = headers.is_a?(Hash) ? headers["content-length"] : nil
      no_body = status == 204 || status == 304 || content_length.to_s == "0"

      json_data = nil
      unless no_body
        jf = VoxgigStruct.getprop(fetched, "json")
        if jf.is_a?(Proc)
          begin
            json_data = jf.call
          rescue StandardError
            # Non-JSON body — leave data nil, keep status/headers.
            json_data = nil
          end
        end
      end

      return {
        "ok" => status >= 200 && status < 300,
        "status" => status,
        "headers" => headers,
        "data" => json_data,
      }
    end

    return {
      "ok" => false,
      "err" => ctx.make_error("direct_invalid", "invalid response type"),
    }
  end


  # Idiomatic facade: client.composition.list / client.composition.load({ "id" => ... })
  def composition
    require_relative 'entity/composition_entity'
    @composition ||= CompositionEntity.new(self, nil)
  end

  # Deprecated: use client.composition instead.
  def Composition(data = nil)
    require_relative 'entity/composition_entity'
    CompositionEntity.new(self, data)
  end


  # Idiomatic facade: client.connection.list / client.connection.load({ "id" => ... })
  def connection
    require_relative 'entity/connection_entity'
    @connection ||= ConnectionEntity.new(self, nil)
  end

  # Deprecated: use client.connection instead.
  def Connection(data = nil)
    require_relative 'entity/connection_entity'
    ConnectionEntity.new(self, data)
  end


  # Idiomatic facade: client.disturbance.list / client.disturbance.load({ "id" => ... })
  def disturbance
    require_relative 'entity/disturbance_entity'
    @disturbance ||= DisturbanceEntity.new(self, nil)
  end

  # Deprecated: use client.disturbance instead.
  def Disturbance(data = nil)
    require_relative 'entity/disturbance_entity'
    DisturbanceEntity.new(self, data)
  end


  # Idiomatic facade: client.liveboard.list / client.liveboard.load({ "id" => ... })
  def liveboard
    require_relative 'entity/liveboard_entity'
    @liveboard ||= LiveboardEntity.new(self, nil)
  end

  # Deprecated: use client.liveboard instead.
  def Liveboard(data = nil)
    require_relative 'entity/liveboard_entity'
    LiveboardEntity.new(self, data)
  end


  # Idiomatic facade: client.log.list / client.log.load({ "id" => ... })
  def log
    require_relative 'entity/log_entity'
    @log ||= LogEntity.new(self, nil)
  end

  # Deprecated: use client.log instead.
  def Log(data = nil)
    require_relative 'entity/log_entity'
    LogEntity.new(self, data)
  end


  # Idiomatic facade: client.occupancy.list / client.occupancy.load({ "id" => ... })
  def occupancy
    require_relative 'entity/occupancy_entity'
    @occupancy ||= OccupancyEntity.new(self, nil)
  end

  # Deprecated: use client.occupancy instead.
  def Occupancy(data = nil)
    require_relative 'entity/occupancy_entity'
    OccupancyEntity.new(self, data)
  end


  # Idiomatic facade: client.station.list / client.station.load({ "id" => ... })
  def station
    require_relative 'entity/station_entity'
    @station ||= StationEntity.new(self, nil)
  end

  # Deprecated: use client.station instead.
  def Station(data = nil)
    require_relative 'entity/station_entity'
    StationEntity.new(self, data)
  end


  # Idiomatic facade: client.vehicle.list / client.vehicle.load({ "id" => ... })
  def vehicle
    require_relative 'entity/vehicle_entity'
    @vehicle ||= VehicleEntity.new(self, nil)
  end

  # Deprecated: use client.vehicle instead.
  def Vehicle(data = nil)
    require_relative 'entity/vehicle_entity'
    VehicleEntity.new(self, data)
  end



  def self.test(testopts = nil, sdkopts = nil)
    sdkopts = sdkopts || {}
    sdkopts = VoxgigStruct.clone(sdkopts)
    sdkopts = {} unless sdkopts.is_a?(Hash)

    testopts = testopts || {}
    testopts = VoxgigStruct.clone(testopts)
    testopts = {} unless testopts.is_a?(Hash)
    testopts["active"] = true

    VoxgigStruct.setpath(sdkopts, "feature.test", testopts)

    sdk = IrailSDK.new(sdkopts)
    sdk.mode = "test"
    sdk
  end
end

File: lua/irail_sdk.lua
-- Irail SDK

local vs = require("utility.struct.struct")
local Utility = require("core.utility_type")
local Spec = require("core.spec")
local helpers = require("core.helpers")

-- Load utility registration (populates Utility._registrar)
require("utility.register")

-- Load features
local BaseFeature = require("feature.base_feature")
local features_factory = require("features")


local IrailSDK = {}
IrailSDK.__index = IrailSDK


local function _make_feature(name)
  local factory = features_factory[name]
  if factory ~= nil then
    return factory()
  end
  return features_factory.base()
end

IrailSDK._make_feature = _make_feature


function IrailSDK.new(options)
  local self = setmetatable({}, IrailSDK)
  self.mode = "live"
  self.features = {}
  self.options = nil

  local utility = Utility.new()
  self._utility = utility

  local config = require("config")()

  self._rootctx = utility.make_context({
    client = self,
    utility = utility,
    config = config,
    options = options or {},
    shared = {},
  }, nil)

  self.options = utility.make_options(self._rootctx)

  if vs.getpath(self.options, "feature.test.active") == true then
    self.mode = "test"
  end

  self._rootctx.options = self.options

  -- Add features from config.
  local feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
  if feature_opts ~= nil then
    local feature_items = vs.items(feature_opts)
    if feature_items ~= nil then
      for _, item in ipairs(feature_items) do
        local fname = item[1]
        local fopts = helpers.to_map(item[2])
        if fopts ~= nil and fopts["active"] == true then
          utility.feature_add(self._rootctx, _make_feature(fname))
        end
      end
    end
  end

  -- Add extension features.
  local extend = vs.getprop(self.options, "extend")
  if type(extend) == "table" then
    for _, f in ipairs(extend) do
      if type(f) == "table" and type(f.get_name) == "function" then
        utility.feature_add(self._rootctx, f)
      end
    end
  end

  -- Initialize features.
  for _, f in ipairs(self.features) do
    utility.feature_init(self._rootctx, f)
  end

  utility.feature_hook(self._rootctx, "PostConstruct")

  -- #BuildFeatures

  return self
end


function IrailSDK:options_map()
  local out = vs.clone(self.options)
  if type(out) == "table" then
    return out
  end
  return {}
end


function IrailSDK:get_utility()
  return Utility.copy(self._utility)
end


function IrailSDK:get_root_ctx()
  return self._rootctx
end


function IrailSDK:prepare(fetchargs)
  local utility = self._utility

  fetchargs = fetchargs or {}

  local ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl")) or {}

  local ctx = utility.make_context({
    opname = "prepare",
    ctrl = ctrl,
  }, self._rootctx)

  local options = self.options

  local path = vs.getprop(fetchargs, "path") or ""
  if type(path) ~= "string" then path = "" end

  local method = vs.getprop(fetchargs, "method") or "GET"
  if type(method) ~= "string" then method = "GET" end

  local params = helpers.to_map(vs.getprop(fetchargs, "params")) or {}
  local query = helpers.to_map(vs.getprop(fetchargs, "query")) or {}

  local headers = utility.prepare_headers(ctx)

  local base = vs.getprop(options, "base") or ""
  if type(base) ~= "string" then base = "" end
  local prefix = vs.getprop(options, "prefix") or ""
  if type(prefix) ~= "string" then prefix = "" end
  local suffix = vs.getprop(options, "suffix") or ""
  if type(suffix) ~= "string" then suffix = "" end

  ctx.spec = Spec.new({
    base = base,
    prefix = prefix,
    suffix = suffix,
    path = path,
    method = method,
    params = params,
    query = query,
    headers = headers,
    body = vs.getprop(fetchargs, "body"),
    step = "start",
  })

  -- Merge user-provided headers.
  local uh = vs.getprop(fetchargs, "headers")
  if type(uh) == "table" then
    for k, v in pairs(uh) do
      ctx.spec.headers[k] = v
    end
  end

  local _, err = utility.prepare_auth(ctx)
  if err ~= nil then
    return nil, err
  end

  return utility.make_fetch_def(ctx)
end


function IrailSDK:direct(fetchargs)
  local utility = self._utility

  local fetchdef, err = self:prepare(fetchargs)
  if err ~= nil then
    return { ok = false, err = err }, nil
  end

  fetchargs = fetchargs or {}
  local ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl")) or {}

  local ctx = utility.make_context({
    opname = "direct",
    ctrl = ctrl,
  }, self._rootctx)

  local url = fetchdef["url"] or ""
  local fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

  if fetch_err ~= nil then
    return { ok = false, err = fetch_err }, nil
  end

  if fetched == nil then
    return {
      ok = false,
      err = ctx:make_error("direct_no_response", "response: undefined"),
    }, nil
  end

  if type(fetched) == "table" then
    local status = helpers.to_int(vs.getprop(fetched, "status"))
    local headers = vs.getprop(fetched, "headers") or {}

    -- No-body responses (204, 304) and explicit zero content-length
    -- must skip JSON parsing — calling json() on an empty body errors.
    local content_length = nil
    if type(headers) == "table" then
      content_length = headers["content-length"]
    end
    local no_body = status == 204 or status == 304 or tostring(content_length) == "0"

    local json_data = nil
    if not no_body then
      local jf = vs.getprop(fetched, "json")
      if type(jf) == "function" then
        local ok, result = pcall(jf)
        if ok then
          json_data = result
        end
        -- Non-JSON body: json_data stays nil, status/headers preserved.
      end
    end

    return {
      ok = status >= 200 and status < 300,
      status = status,
      headers = headers,
      data = json_data,
    }, nil
  end

  return {
    ok = false,
    err = ctx:make_error("direct_invalid", "invalid response type"),
  }, nil
end



-- Idiomatic facade: client:Composition():list() / client:Composition():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function IrailSDK:Composition(data)
  local EntityMod = require("entity.composition_entity")
  if data == nil then
    if self._composition == nil then
      self._composition = EntityMod.new(self, nil)
    end
    return self._composition
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Connection():list() / client:Connection():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function IrailSDK:Connection(data)
  local EntityMod = require("entity.connection_entity")
  if data == nil then
    if self._connection == nil then
      self._connection = EntityMod.new(self, nil)
    end
    return self._connection
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Disturbance():list() / client:Disturbance():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function IrailSDK:Disturbance(data)
  local EntityMod = require("entity.disturbance_entity")
  if data == nil then
    if self._disturbance == nil then
      self._disturbance = EntityMod.new(self, nil)
    end
    return self._disturbance
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Liveboard():list() / client:Liveboard():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function IrailSDK:Liveboard(data)
  local EntityMod = require("entity.liveboard_entity")
  if data == nil then
    if self._liveboard == nil then
      self._liveboard = EntityMod.new(self, nil)
    end
    return self._liveboard
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Log():list() / client:Log():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function IrailSDK:Log(data)
  local EntityMod = require("entity.log_entity")
  if data == nil then
    if self._log == nil then
      self._log = EntityMod.new(self, nil)
    end
    return self._log
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Occupancy():list() / client:Occupancy():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function IrailSDK:Occupancy(data)
  local EntityMod = require("entity.occupancy_entity")
  if data == nil then
    if self._occupancy == nil then
      self._occupancy = EntityMod.new(self, nil)
    end
    return self._occupancy
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Station():list() / client:Station():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function IrailSDK:Station(data)
  local EntityMod = require("entity.station_entity")
  if data == nil then
    if self._station == nil then
      self._station = EntityMod.new(self, nil)
    end
    return self._station
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Vehicle():list() / client:Vehicle():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function IrailSDK:Vehicle(data)
  local EntityMod = require("entity.vehicle_entity")
  if data == nil then
    if self._vehicle == nil then
      self._vehicle = EntityMod.new(self, nil)
    end
    return self._vehicle
  end
  return EntityMod.new(self, data)
end




function IrailSDK.test(testopts, sdkopts)
  sdkopts = sdkopts or {}
  sdkopts = vs.clone(sdkopts)
  if type(sdkopts) ~= "table" then
    sdkopts = {}
  end

  testopts = testopts or {}
  testopts = vs.clone(testopts)
  if type(testopts) ~= "table" then
    testopts = {}
  end
  testopts["active"] = true

  vs.setpath(sdkopts, "feature.test", testopts)

  local sdk = IrailSDK.new(sdkopts)
  sdk.mode = "test"

  return sdk
end


return IrailSDK
